src/Compilers/Core/Rebuild/BuildConstructor.cs

```diff
@@ -349,6 +270,66 @@ private static VisualBasicCompilationOptions CreateVisualBasicCompilationOptions
         }
-        #endregion
+
+        public static unsafe EmitResult Emit(
+            Stream rebuildPeStream,
+            FileInfo originalBinaryPath,
+            CompilationOptionsReader optionsReader,
+            Compilation producedCompilation,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var peHeader = optionsReader.PeReader.PEHeaders.PEHeader!;
+            var win32Resources = optionsReader.PeReader.GetSectionData(peHeader.ResourceTableDirectory.RelativeVirtualAddress);
+            using var win32ResourceStream = new UnmanagedMemoryStream(win32Resources.Pointer, win32Resources.Length);
+
+            var sourceLink = optionsReader.GetSourceLinkUTF8();
+
+            var embeddedTexts = producedCompilation.SyntaxTrees
+                    .Select(st => (path: st.FilePath, text: st.GetText()))
+                    .Where(pair => pair.text.CanBeEmbedded)
+                    .Select(pair => EmbeddedText.FromSource(pair.path, pair.text))
+                    .ToImmutableArray();
+
+            var debugEntryPoint = getDebugEntryPoint();
+
+            var emitResult = producedCompilation.Emit(
+                peStream: rebuildPeStream,
+                pdbStream: null,
+                xmlDocumentationStream: null,
+                win32Resources: win32ResourceStream,
+                useRawWin32Resources: true,
+                manifestResources: optionsReader.GetManifestResources(),
+                options: new EmitOptions(
+                    debugInformationFormat: DebugInformationFormat.Embedded,
+                    highEntropyVirtualAddressSpace: (peHeader.DllCharacteristics & DllCharacteristics.HighEntropyVirtualAddressSpace) != 0,
+                    subsystemVersion: SubsystemVersion.Create(peHeader.MajorSubsystemVersion, peHeader.MinorSubsystemVersion)),
+                debugEntryPoint: debugEntryPoint,
+                metadataPEStream: null,
+                pdbOptionsBlobReader: optionsReader.GetMetadataCompilationOptionsBlobReader(),
+                sourceLinkStream: sourceLink != null ? new MemoryStream(sourceLink) : null,
+                embeddedTexts: embeddedTexts,
+                cancellationToken: cancellationToken);
+
+            return emitResult;
+
+            IMethodSymbol? getDebugEntryPoint()
+            {
+                if (optionsReader.GetMainTypeName() is { } mainTypeName &&
+                    optionsReader.GetMainMethodName() is { } mainMethodName)
+                {
+                    var typeSymbol = producedCompilation.GetTypeByMetadataName(mainTypeName);
+                    if (typeSymbol is object)
+                    {
+                        var methodSymbols = typeSymbol
+                            .GetMembers(mainMethodName)
+                            .OfType<IMethodSymbol>();
+                        return methodSymbols.FirstOrDefault();
```

Comment:
>FirstOrDefault [](start = 45, length = 14)

Do we need to handle the case where there are multiple "Main" methods in the type? Could we simply use `producedCompilation.GetEntryPoint()` instead (and avoid looking at `optionsReader` as well)?

Reply:
I'm interested in looking more closely at this, as I believe it is buggy, however in this PR I will just migrate this as-is.